src/Orleans/Messaging/IncomingMessageBuffer.cs

```diff
@@ -1,44 +1,329 @@
+                result,
+                readBuffer,
+                decodeOffset,
+                length,
+                out bytesRead);
+
+            // adjust the read cursor
+            decodeOffset += bytesRead;
+            return result;
+        }
+
+        // Hold parts of incomplete message between recieve calls
+        private class MessagePrefixHolder
+        {
+            private List<ArraySegment<byte>> buf;
+            public bool HasPrefix { get; private set; }
+            public int Count { get; private set; }
+
+            public void HandlePrefix(List<ArraySegment<byte>> buffer, int offset, int remainingBytesToProcess)
+            {
+                if (remainingBytesToProcess == 0)
+                    return;
+                HasPrefix = true;
+                Count += remainingBytesToProcess;
+                var length = remainingBytesToProcess;
+                buf = buf ?? new List<ArraySegment<byte>>();
+                var lengthSoFar = 0;
+                var countSoFar = 0;
+                foreach (var segment in buffer)
+                {
+                    var bytesStillToSkip = offset - lengthSoFar;
+                    lengthSoFar += segment.Count;
+
+                    if (segment.Count <= bytesStillToSkip) // Still skipping past this buffer
+                    {
+                        continue;
+                    }
+                    if (bytesStillToSkip > 0) // This is the first buffer
+                    {
+                        var count = Math.Min(length - countSoFar, segment.Count - bytesStillToSkip);
+
+                        // Not quite efficient, todo: use local circular buffer instead
+                        var seg = new ArraySegment<byte>(BufferPool.GlobalPool.GetBuffer(), 0, count);
+
+                        // Maintaining of the ownership is important, as the source buffer will be refilled after next receive
+                        Buffer.BlockCopy(segment.Array, bytesStillToSkip, seg.Array, 0, count);
+                        buf.Add(seg);
+                        countSoFar += count;
+                    }
+                    else
+                    {
+                        var count = Math.Min(length - countSoFar, segment.Count);
+                        var seg = new ArraySegment<byte>(BufferPool.GlobalPool.GetBuffer(), 0, count);
+
+                        Buffer.BlockCopy(segment.Array, 0, seg.Array, 0, count);
+                        buf.Add(seg);
+                        countSoFar += count;
+                    }
+
+                    if (countSoFar == length)
+                    {
+                        break;
+                    }
+                }
+            }
+
+            public List<ArraySegment<byte>> GetPrefix()
+            {
+                return buf;
+            }
+
+            public void Reset()
+            {
+                Count = 0;
+                HasPrefix = false;
+                BufferPool.GlobalPool.Release(buf);
```

Comment:
BuildSegmentListWithLengthLimit does not copy the data, it just carves up the list of segments.  This means we're returning buffer to the pool that are still being used in the message we are currently processing.